tiledb/sm/cpp_api/context.h

```diff
@@ -103,6 +103,26 @@ class Context {
     set_tag("x-tiledb-api-language", "c++");
   }
 
+  /**
+   * Constructor. Creates a TileDB context with an optional configuration.
+   * @throws TileDBError if construction fails
+   */
+  explicit Context(std::optional<Config> config) {
```

Comment:
The type `optional<Config>` means the `config` argument is passed by value, meaning it will be copied. That's not what we want for large objects like `Config`.

In order to combine the constructors for both empty and non-empty cases, this should be `optional<const Config&>`. The problem is that `optional<T&>` won't compile. From https://en.cppreference.com/w/cpp/utility/optional
>There are no optional references; a program is ill-formed if it instantiates an optional with a reference type.

The problem is that the origin of this story is asking for pointer semantics for constructors that use references. That's best done outside this code entirely.